File: server/schemas.py
"""
Pydantic Schemas
================

Request/Response models for the API endpoints.
"""

import base64
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

# ============================================================================
# Project Schemas
# ============================================================================

class ProjectCreate(BaseModel):
    """Request schema for creating a new project."""
    name: str = Field(..., min_length=1, max_length=50, pattern=r'^[a-zA-Z0-9_-]+$')
    git_url: str = Field(..., min_length=1, description="Git repository URL (https:// or git@)")
    is_new: bool = Field(default=True, description="True if this is a new project needing wizard setup")
    spec_method: Literal["claude", "manual"] = "claude"


class ProjectStats(BaseModel):
    """Project statistics."""
    passing: int = 0
    in_progress: int = 0
    total: int = 0
    percentage: float = 0.0


class ProjectSummary(BaseModel):
    """Summary of a project for list view."""
    name: str
    git_url: str
    local_path: str
    is_new: bool = True
    has_spec: bool
    wizard_incomplete: bool = False
    stats: ProjectStats
    target_container_count: int = 1
    agent_status: str | None = None
    agent_running: bool | None = None
    agent_model: str | None = None  # Model for coder/overseer agents


class ProjectDetail(BaseModel):
    """Detailed project information."""
    name: str
    git_url: str
    local_path: str
    is_new: bool = True
    has_spec: bool
    stats: ProjectStats
    prompts_dir: str
    target_container_count: int = 1
    agent_model: str | None = None  # Model for coder/overseer agents


class ProjectSettingsUpdate(BaseModel):
    """Request schema for updating project settings."""
    agent_model: str = Field(..., description="Model ID for coder/overseer agents")


class ProjectPrompts(BaseModel):
    """Project prompt files content."""
    app_spec: str = ""
    initializer_prompt: str = ""
    coding_prompt: str = ""


class ProjectPromptsUpdate(BaseModel):
    """Request schema for updating project prompts."""
    app_spec: str | None = None
    initializer_prompt: str | None = None
    coding_prompt: str | None = None


class WizardStatusMessage(BaseModel):
    """A chat message stored in wizard status."""
    role: Literal["user", "assistant"]
    content: str
    timestamp: datetime


class WizardStatus(BaseModel):
    """Wizard state for resuming interrupted project setup."""
    step: Literal["name", "folder", "method", "chat"]
    spec_method: Literal["claude", "manual"] | None = None
    started_at: datetime
    chat_messages: list[WizardStatusMessage] = []


class AddExistingRepoRequest(BaseModel):
    """Request schema for adding an existing repository."""
    name: str = Field(..., min_length=1, max_length=50, pattern=r'^[a-zA-Z0-9_-]+$')
    git_url: str = Field(..., min_length=1, description="Git repository URL (https:// or git@)")


class ContainerCountUpdate(BaseModel):
    """Request to update target container count."""
    target_count: int = Field(..., ge=1, le=10)


# ============================================================================
# Container Schemas
# ============================================================================

class ContainerStatus(BaseModel):
    """Container instance status."""
    id: int
    container_number: int
    container_type: Literal["init", "coding"]
    status: Literal["created", "running", "stopping", "stopped"]
    current_feature: str | None = None
    docker_container_id: str | None = None


# ============================================================================
# Feature Schemas
# ============================================================================

class FeatureBase(BaseModel):
    """Base feature attributes."""
    category: str
    name: str
    description: str
    steps: list[str]


class FeatureCreate(FeatureBase):
    """Request schema for creating a new feature."""
    priority: int | None = None


class FeatureUpdate(BaseModel):
    """Request schema for updating a feature."""
    name: str | None = None
    description: str | None = None
    category: str | None = None
    priority: int | None = None
    steps: list[str] | None = None


class FeatureResponse(FeatureBase):
    """Response schema for a feature."""
    id: str  # beads uses string IDs like "feat-1"
    priority: int
    passes: bool
    in_progress: bool

    class Config:
        from_attributes = True


class FeatureListResponse(BaseModel):
    """Response containing list of features organized by status."""
    pending: list[FeatureResponse]
    in_progress: list[FeatureResponse]
    done: list[FeatureResponse]


# ============================================================================
# Agent Schemas
# ============================================================================

class AgentStartRequest(BaseModel):
    """Request schema for starting the agent."""
    instruction: str | None = None  # Instruction to send to Claude Code
    yolo_mode: bool = False  # Kept for backwards compatibility


class AgentStatus(BaseModel):
    """Current agent/container status."""
    status: Literal["not_created", "stopped", "running", "paused", "crashed", "completed"]
    container_name: str | None = None
    started_at: datetime | None = None
    idle_seconds: int = 0
    agent_running: bool = False  # True if agent process is running inside container
    graceful_stop_requested: bool = False  # True if graceful stop has been requested
    # Legacy fields for backwards compatibility
    pid: int | None = None
    yolo_mode: bool = False


class AgentActionResponse(BaseModel):
    """Response for agent control actions."""
    success: bool
    status: str
    message: str = ""


# ============================================================================
# Setup Schemas
# ============================================================================

class SetupStatus(BaseModel):
    """System setup status."""
    claude_cli: bool
    credentials: bool
    node: bool
    npm: bool


# ============================================================================
# WebSocket Message Schemas
# ============================================================================

class WSProgressMessage(BaseModel):
    """WebSocket message for progress updates."""
    type: Literal["progress"] = "progress"
    passing: int
    total: int
    percentage: float


class WSFeatureUpdateMessage(BaseModel):
    """WebSocket message for feature status updates."""
    type: Literal["feature_update"] = "feature_update"
    feature_id: str  # beads uses string IDs
    passes: bool


class WSLogMessage(BaseModel):
    """WebSocket message for agent log output."""
    type: Literal["log"] = "log"
    line: str
    timestamp: datetime


class WSAgentStatusMessage(BaseModel):
    """WebSocket message for agent status changes."""
    type: Literal["agent_status"] = "agent_status"
    status: str


# ============================================================================
# Spec Chat Schemas
# ============================================================================

# Maximum file sizes
MAX_IMAGE_SIZE = 5 * 1024 * 1024  # 5 MB for images
MAX_TEXT_SIZE = 1 * 1024 * 1024   # 1 MB for text files

# Supported MIME types
IMAGE_MIME_TYPES = Literal['image/jpeg', 'image/png']
TEXT_MIME_TYPES = Literal[
    'text/plain', 'text/markdown', 'text/csv', 'application/json',
    'text/html', 'text/css', 'text/javascript', 'application/xml'
]


class ImageAttachment(BaseModel):
    """Image attachment from client for spec creation chat."""
    filename: str = Field(..., min_length=1, max_length=255)
    mimeType: IMAGE_MIME_TYPES
    base64Data: str
    isText: Literal[False] = False

    @field_validator('base64Data')
    @classmethod
    def validate_base64_and_size(cls, v: str) -> str:
        """Validate that base64 data is valid and within size limit."""
        try:
            decoded = base64.b64decode(v)
            if len(decoded) > MAX_IMAGE_SIZE:
                raise ValueError(
                    f'Image size ({len(decoded) / (1024 * 1024):.1f} MB) exceeds '
                    f'maximum of {MAX_IMAGE_SIZE // (1024 * 1024)} MB'
                )
            return v
        except Exception as e:
            if 'Image size' in str(e):
                raise
            raise ValueError(f'Invalid base64 data: {e}')


class TextAttachment(BaseModel):
    """Text file attachment from client for spec creation chat."""
    filename: str = Field(..., min_length=1, max_length=255)
    mimeType: TEXT_MIME_TYPES
    textContent: str
    isText: Literal[True] = True

    @field_validator('textContent')
    @classmethod
    def validate_text_size(cls, v: str) -> str:
        """Validate that text content is within size limit."""
        size = len(v.encode('utf-8'))
        if size > MAX_TEXT_SIZE:
            raise ValueError(
                f'Text file size ({size / (1024 * 1024):.1f} MB) exceeds '
                f'maximum of {MAX_TEXT_SIZE // (1024 * 1024)} MB'
            )
        return v


# Union type for any attachment
FileAttachment = ImageAttachment | TextAttachment


# ============================================================================
# Task Schemas (Edit Mode)
# ============================================================================

class TaskCreate(BaseModel):
    """Request schema for creating a task in edit mode."""
    title: str = Field(..., min_length=1, max_length=200)
    description: str = Field(default="", max_length=5000)
    priority: int = Field(default=2, ge=0, le=4)
    task_type: str = Field(default="feature", pattern=r'^(feature|task|bug)$')


class TaskUpdate(BaseModel):
    """Request schema for updating a task in edit mode."""
    status: str | None = Field(default=None, pattern=r'^(open|in_progress|closed)$')
    priority: int | None = Field(default=None, ge=0, le=4)
    title: str | None = Field(default=None, min_length=1, max_length=200)

File: packages/shared/src/schemas.ts
/**
 * Shared Zod schemas for ZeroCoder
 * Matches server/schemas.py Pydantic models
 */

import { z } from 'zod';

// ============================================================================
// Constants
// ============================================================================

export const MAX_IMAGE_SIZE = 5 * 1024 * 1024; // 5 MB for images
export const MAX_TEXT_SIZE = 1 * 1024 * 1024; // 1 MB for text files

// ============================================================================
// Project Schemas
// ============================================================================

export const ProjectCreateSchema = z.object({
  name: z
    .string()
    .min(1)
    .max(50)
    .regex(/^[a-zA-Z0-9_-]+$/),
  git_url: z.string().min(1).describe('Git repository URL (https:// or git@)'),
  is_new: z.boolean().default(true).describe('True if this is a new project needing wizard setup'),
  spec_method: z.enum(['claude', 'manual']).default('claude'),
});

export const ProjectStatsSchema = z.object({
  passing: z.number().int().default(0),
  in_progress: z.number().int().default(0),
  total: z.number().int().default(0),
  percentage: z.number().default(0.0),
});

export const ProjectSummarySchema = z.object({
  name: z.string(),
  git_url: z.string(),
  local_path: z.string(),
  is_new: z.boolean().default(true),
  has_spec: z.boolean(),
  wizard_incomplete: z.boolean().default(false),
  stats: ProjectStatsSchema,
  target_container_count: z.number().int().default(1),
  agent_status: z.string().nullable().optional(),
  agent_running: z.boolean().nullable().optional(),
  agent_model: z.string().nullable().optional(),
});

export const ProjectDetailSchema = z.object({
  name: z.string(),
  git_url: z.string(),
  local_path: z.string(),
  is_new: z.boolean().default(true),
  has_spec: z.boolean(),
  stats: ProjectStatsSchema,
  prompts_dir: z.string(),
  target_container_count: z.number().int().default(1),
  agent_model: z.string().nullable().optional(),
});

export const ProjectSettingsUpdateSchema = z.object({
  agent_model: z.string().describe('Model ID for coder/overseer agents'),
});

export const ProjectPromptsSchema = z.object({
  app_spec: z.string().default(''),
  initializer_prompt: z.string().default(''),
  coding_prompt: z.string().default(''),
});

export const ProjectPromptsUpdateSchema = z.object({
  app_spec: z.string().nullable().optional(),
  initializer_prompt: z.string().nullable().optional(),
  coding_prompt: z.string().nullable().optional(),
});

export const WizardStatusMessageSchema = z.object({
  role: z.enum(['user', 'assistant']),
  content: z.string(),
  timestamp: z.string(), // ISO datetime string
});

export const WizardStatusSchema = z.object({
  step: z.enum(['mode', 'details', 'method', 'chat']),
  spec_method: z.enum(['claude', 'manual']).nullable().optional(),
  started_at: z.string(), // ISO datetime string
  chat_messages: z.array(WizardStatusMessageSchema).default([]),
});

export const AddExistingRepoRequestSchema = z.object({
  name: z
    .string()
    .min(1)
    .max(50)
    .regex(/^[a-zA-Z0-9_-]+$/),
  git_url: z.string().min(1).describe('Git repository URL (https:// or git@)'),
});

export const ContainerCountUpdateSchema = z.object({
  target_count: z.number().int().min(1).max(10),
});

// ============================================================================
// Container Schemas
// ============================================================================

export const ContainerTypeSchema = z.enum(['init', 'coding']);
export const ContainerStatusTypeSchema = z.enum([
  'not_created',
  'created',
  'running',
  'stopping',
  'stopped',
  'completed',
]);
export const AgentTypeSchema = z.enum(['coder', 'overseer', 'initializer', 'reviewer']);
export const SdkTypeSchema = z.enum(['claude', 'opencode']);

export const ContainerStatusSchema = z.object({
  id: z.number().int(),
  container_number: z.number().int(),
  container_type: ContainerTypeSchema,
  status: ContainerStatusTypeSchema,
  current_feature: z.string().nullable().optional(),
  docker_container_id: z.string().nullable().optional(),
  agent_type: AgentTypeSchema.nullable().optional(),
  sdk_type: SdkTypeSchema.nullable().optional(),
});

// ============================================================================
// Feature Schemas
// ============================================================================

export const FeatureBaseSchema = z.object({
  category: z.string(),
  name: z.string(),
  description: z.string(),
  steps: z.array(z.string()),
});

export const FeatureCreateSchema = FeatureBaseSchema.extend({
  priority: z.number().int().nullable().optional(),
});

export const FeatureUpdateSchema = z.object({
  name: z.string().nullable().optional(),
  description: z.string().nullable().optional(),
  category: z.string().nullable().optional(),
  priority: z.number().int().nullable().optional(),
  steps: z.array(z.string()).nullable().optional(),
});

export const FeatureResponseSchema = FeatureBaseSchema.extend({
  id: z.string(), // beads uses string IDs like "feat-1"
  priority: z.number().int(),
  passes: z.boolean(),
  in_progress: z.boolean(),
});

export const FeatureListResponseSchema = z.object({
  pending: z.array(FeatureResponseSchema),
  in_progress: z.array(FeatureResponseSchema),
  done: z.array(FeatureResponseSchema),
});

// ============================================================================
// Agent Schemas
// ============================================================================

export const AgentStatusTypeSchema = z.enum([
  'not_created',
  'stopped',
  'running',
  'paused',
  'crashed',
  'completed',
]);

export const AgentStartRequestSchema = z.object({
  instruction: z.string().nullable().optional(),
  yolo_mode: z.boolean().default(false), // Kept for backwards compatibility
});

export const AgentStatusSchema = z.object({
  status: AgentStatusTypeSchema,
  container_name: z.string().nullable().optional(),
  started_at: z.string().nullable().optional(), // ISO datetime string
  idle_seconds: z.number().int().default(0),
  agent_running: z.boolean().default(false),
  graceful_stop_requested: z.boolean().default(false),
  current_feature: z.string().nullable().optional(),
  agent_type: AgentTypeSchema.nullable().optional(),
  sdk_type: SdkTypeSchema.nullable().optional(),
  // Legacy fields for backwards compatibility
  pid: z.number().int().nullable().optional(),
  yolo_mode: z.boolean().default(false),
});

export const AgentActionResponseSchema = z.object({
  success: z.boolean(),
  status: z.string(),
  message: z.string().default(''),
});

// ============================================================================
// Setup Schemas
// ============================================================================

export const SetupStatusSchema = z.object({
  claude_cli: z.boolean(),
  credentials: z.boolean(),
  node: z.boolean(),
  npm: z.boolean(),
});

// ============================================================================
// WebSocket Message Schemas
// ============================================================================

export const WSProgressMessageSchema = z.object({
  type: z.literal('progress'),
  passing: z.number().int(),
  total: z.number().int(),
  percentage: z.number(),
});

export const WSFeatureUpdateMessageSchema = z.object({
  type: z.literal('feature_update'),
  feature_id: z.string(), // beads uses string IDs
  passes: z.boolean(),
});

export const WSLogMessageSchema = z.object({
  type: z.literal('log'),
  line: z.string(),
  timestamp: z.string(), // ISO datetime string
});

export const WSAgentStatusMessageSchema = z.object({
  type: z.literal('agent_status'),
  status: z.string(),
});

// ============================================================================
// Spec Chat / Attachment Schemas
// ============================================================================

export const ImageMimeTypeSchema = z.enum(['image/jpeg', 'image/png']);
export const TextMimeTypeSchema = z.enum([
  'text/plain',
  'text/markdown',
  'text/csv',
  'application/json',
  'text/html',
  'text/css',
  'text/javascript',
  'application/xml',
]);

export const ImageAttachmentSchema = z.object({
  filename: z.string().min(1).max(255),
  mimeType: ImageMimeTypeSchema,
  base64Data: z.string().refine(
    (data) => {
      try {
        const decoded = atob(data);
        return decoded.length <= MAX_IMAGE_SIZE;
      } catch {
        return false;
      }
    },
    { message: `Image must be valid base64 and under ${MAX_IMAGE_SIZE / (1024 * 1024)} MB` }
  ),
  isText: z.literal(false).default(false),
});

export const TextAttachmentSchema = z.object({
  filename: z.string().min(1).max(255),
  mimeType: TextMimeTypeSchema,
  textContent: z.string().refine((text) => Buffer.byteLength(text, 'utf8') <= MAX_TEXT_SIZE, {
    message: `Text file must be under ${MAX_TEXT_SIZE / (1024 * 1024)} MB`,
  }),
  isText: z.literal(true).default(true),
});

export const FileAttachmentSchema = z.discriminatedUnion('isText', [
  ImageAttachmentSchema.extend({ isText: z.literal(false) }),
  TextAttachmentSchema.extend({ isText: z.literal(true) }),
]);

// ============================================================================
// Task Schemas (Edit Mode)
// ============================================================================

export const TaskCreateSchema = z.object({
  title: z.string().min(1).max(200),
  description: z.string().max(5000).default(''),
  priority: z.number().int().min(0).max(4).default(2),
  task_type: z.enum(['feature', 'task', 'bug']).default('feature'),
});

export const TaskUpdateSchema = z.object({
  status: z.enum(['open', 'in_progress', 'closed']).nullable().optional(),
  priority: z.number().int().min(0).max(4).nullable().optional(),
  title: z.string().min(1).max(200).nullable().optional(),
});

// ============================================================================
// Remote Machine Schemas
// ============================================================================

export const RemoteMachineCreateSchema = z.object({
  name: z.string().min(1).max(100),
  host: z.string().min(1).max(255),
  port: z.number().int().min(1).max(65535).default(22),
  username: z.string().min(1).max(100).default('root'),
  ssh_key_path: z.string().max(500).nullable().optional(),
  git_ssh_key_path: z.string().max(500).nullable().optional(), // Path to SSH key for git clone on remote
});

export const RemoteMachineResponseSchema = z.object({
  id: z.number().int(),
  name: z.string(),
  host: z.string(),
  port: z.number().int(),
  username: z.string(),
  ssh_key_path: z.string().nullable().optional(),
  git_ssh_key_path: z.string().nullable().optional(),
  status: z.string(),
  last_checked_at: z.string().nullable().optional(),
  created_at: z.string().nullable().optional(),
  // Daemon fields
  daemon_port: z.number().int().nullable().optional(),
  daemon_pid: z.number().int().nullable().optional(),
  daemon_last_seen: z.string().nullable().optional(),
});

export const RemoteAgentStartRequestSchema = z.object({
  machine_id: z.number().int(),
});

export const RemoteAgentStatusResponseSchema = z.object({
  id: z.number().int(),
  project_name: z.string(),
  machine_id: z.number().int(),
  machine_name: z.string(),
  agent_number: z.number().int(),
  status: z.string(),
  current_feature: z.string().nullable().optional(),
  pid: z.number().int().nullable().optional(),
  graceful_stop_requested: z.boolean(),
  restarting: z.boolean(),
  last_activity_at: z.string().nullable().optional(),
});

// ============================================================================
// Inferred Types (types not defined in types.ts)
// ============================================================================

// These types are only exported from schemas because they're not in types.ts
export type ProjectCreate = z.infer<typeof ProjectCreateSchema>;
export type ProjectSettingsUpdate = z.infer<typeof ProjectSettingsUpdateSchema>;
export type ProjectPromptsUpdate = z.infer<typeof ProjectPromptsUpdateSchema>;
export type AddExistingRepoRequest = z.infer<typeof AddExistingRepoRequestSchema>;
export type ContainerCountUpdate = z.infer<typeof ContainerCountUpdateSchema>;
export type ContainerStatus = z.infer<typeof ContainerStatusSchema>;
export type FeatureBase = z.infer<typeof FeatureBaseSchema>;
export type FeatureUpdate = z.infer<typeof FeatureUpdateSchema>;
export type FeatureResponse = z.infer<typeof FeatureResponseSchema>;
export type AgentStatusType = z.infer<typeof AgentStatusTypeSchema>;
export type AgentStartRequest = z.infer<typeof AgentStartRequestSchema>;
export type TaskCreate = z.infer<typeof TaskCreateSchema>;
export type TaskUpdate = z.infer<typeof TaskUpdateSchema>;
export type RemoteMachineResponse = z.infer<typeof RemoteMachineResponseSchema>;
export type RemoteAgentStartRequest = z.infer<typeof RemoteAgentStartRequestSchema>;
export type RemoteAgentStatusResponse = z.infer<typeof RemoteAgentStatusResponseSchema>;
